File: resources/js/ode/julia/core/controller.js
/* ==========================================================================
   ode_julia / core / controller

   Choosing the next step size from the error estimate.

   The PI controller is OrdinaryDiffEq's default for these methods and is what
   is implemented here, with its coefficients:

       β₁ = 7/(10k),  β₂ = 2/(5k)      k the order of the error estimate
       γ  = 9/10,     qmin = 1/5,      qmax = 10
       qoldinit = 1e-4

   The proportional term alone (the I controller, β₂ = 0) is the textbook
   h·(1/EEst)^(1/k). The integral term damps the oscillation that the plain
   version falls into on stiff problems, where one lucky step is followed by
   one that is far too large. Setting `beta2: 0` recovers the I controller,
   which is what the BDF methods use when they change order.
   ========================================================================== */

/**
 * @param {object} opts
 * @param {number} opts.order      the order of the *error estimate*
 * @param {number} [opts.beta1]
 * @param {number} [opts.beta2]
 * @param {number} [opts.gamma=0.9]      safety factor
 * @param {number} [opts.qmin=0.2]       fastest allowed shrink is dt*qmin
 * @param {number} [opts.qmax=10]        fastest allowed growth is dt*qmax
 * @param {number} [opts.qsteadyMin=1]   inside [min,max] the step is left alone
 * @param {number} [opts.qsteadyMax=1.2]
 * @param {number} [opts.qoldinit=1e-4]
 */
export class PIController {
  constructor(opts) {
    const k = opts.order;
    this.beta1 = opts.beta1 ?? 7 / (10 * k);
    this.beta2 = opts.beta2 ?? 2 / (5 * k);
    this.gamma = opts.gamma ?? 0.9;
    this.qmin = opts.qmin ?? 0.2;
    this.qmax = opts.qmax ?? 10;
    this.qsteadyMin = opts.qsteadyMin ?? 1;
    this.qsteadyMax = opts.qsteadyMax ?? 1.2;
    this.qoldinit = opts.qoldinit ?? 1e-4;
    this.errold = this.qoldinit;
    this.q11 = 1;
  }

  /** Order changed under us (the BDF methods do this); refit the exponents. */
  setOrder(k) {
    this.beta1 = 7 / (10 * k);
    this.beta2 = 2 / (5 * k);
  }

  /**
   * q, the factor the step is *divided* by: dtnew = dt / q.
   * @param {number} EEst  the scaled error estimate; <= 1 is an accepted step
   */
  q(EEst) {
    if (EEst === 0) return 1 / this.qmax;
    this.q11 = EEst ** this.beta1;
    let q = this.q11 / this.errold ** this.beta2;
    q /= this.gamma;
    return Math.min(1 / this.qmin, Math.max(1 / this.qmax, q));
  }

  /** dtnew after an accepted step, and the state update that goes with it. */
  accept(EEst, dt) {
    let q = this.q(EEst);
    if (q >= this.qsteadyMin && q <= this.qsteadyMax) q = 1;
    this.errold = Math.max(EEst, this.qoldinit);
    return dt / q;
  }

  /**
   * dtnew after a rejected step. The proportional part only: the integral term
   * remembers accepted steps, and a rejected one says nothing about the trend.
   *
   * EEst has to be passed in. Reading the q11 left over from the last accepted
   * step instead -- which is what this did at first, and what it looks like
   * OrdinaryDiffEq does until you notice it recomputes the controller on every
   * step, accepted or not -- makes the shrink factor a number about a step
   * that already succeeded. It is then always close to 1, the step comes down
   * by a tenth at a time, and a solver meeting a sharp transition takes tens
   * of thousands of rejections to get through it instead of five.
   */
  reject(EEst, dt) {
    this.q11 = EEst ** this.beta1;
    return dt / Math.min(1 / this.qmin, this.q11 / this.gamma);
  }

  reset() {
    this.errold = this.qoldinit;
    this.q11 = 1;
  }
}

/**
 * Hairer's starting step (Solving ODEs I, II.4), which is what SciML uses.
 *
 * Two explicit Euler probes: the first sizes h from ‖u₀‖/‖f₀‖, the second
 * refines it from the observed second derivative. Cheap -- two extra f calls --
 * and much better than any fixed guess, which on a stiff problem is either
 * thrown away in a dozen rejections or starts so small the first decade of the
 * solution is wasted.
 *
 * @returns {number} a signed step, in the direction of tdir
 */
export function initialStep(f, t0, u0, f0, tdir, order, reltol, abstol, dtmax, work) {
  const n = u0.length;
  const { u1, f1, w } = work;
  const scalarAtol = typeof abstol === 'number';
  for (let i = 0; i < n; i++) w[i] = (scalarAtol ? abstol : abstol[i]) + reltol * Math.abs(u0[i]);

  let d0 = 0;
  let d1 = 0;
  for (let i = 0; i < n; i++) {
    const a = u0[i] / w[i];
    const b = f0[i] / w[i];
    d0 += a * a;
    d1 += b * b;
  }
  d0 = Math.sqrt(d0 / n);
  d1 = Math.sqrt(d1 / n);

  let h0 = (d0 < 1e-5 || d1 < 1e-5) ? 1e-6 : 0.01 * (d0 / d1);
  h0 = Math.min(h0, Math.abs(dtmax));

  for (let i = 0; i < n; i++) u1[i] = u0[i] + tdir * h0 * f0[i];
  f(t0 + tdir * h0, u1, f1);

  let d2 = 0;
  for (let i = 0; i < n; i++) {
    const a = (f1[i] - f0[i]) / w[i];
    d2 += a * a;
  }
  d2 = Math.sqrt(d2 / n) / h0;

  const dmax = Math.max(d1, d2);
  const h1 = dmax <= 1e-15
    ? Math.max(1e-6, h0 * 1e-3)
    : (0.01 / dmax) ** (1 / (order + 1));

  const h = Math.min(100 * h0, h1, Math.abs(dtmax));
  return tdir * h;
}

File: kompartment/python/kompartment/engine/solvers/julia/controller.py
"""The next step size from the error estimate (``core/controller.js``).

OrdinaryDiffEq's PI controller -- beta1 = 7/(10k), beta2 = 2/(5k), gamma 9/10,
qmin 1/5, qmax 10, qoldinit 1e-4 -- and Hairer's starting step.
"""

from __future__ import annotations

import math
from typing import Any, Callable, Optional

import numpy as np

from ._js import jdiv, jmax, jmin, jpow, seq_sum


class PIController:
    """``PIController``: ``accept`` and ``reject`` answer the next step."""

    def __init__(self, order: float, beta1: Optional[float] = None, beta2: Optional[float] = None,
                 gamma: Optional[float] = None, qmin: Optional[float] = None, qmax: Optional[float] = None,
                 qsteady_min: Optional[float] = None, qsteady_max: Optional[float] = None,
                 qoldinit: Optional[float] = None) -> None:
        k = order
        self.beta1 = beta1 if beta1 is not None else 7 / (10 * k)
        self.beta2 = beta2 if beta2 is not None else 2 / (5 * k)
        self.gamma = gamma if gamma is not None else 0.9
        self.qmin = qmin if qmin is not None else 0.2
        self.qmax = qmax if qmax is not None else 10
        self.qsteady_min = qsteady_min if qsteady_min is not None else 1
        self.qsteady_max = qsteady_max if qsteady_max is not None else 1.2
        self.qoldinit = qoldinit if qoldinit is not None else 1e-4
        self.errold = self.qoldinit
        self.q11 = 1.0

    def set_order(self, k: float) -> None:
        self.beta1 = 7 / (10 * k)
        self.beta2 = 2 / (5 * k)

    def q(self, eest: float) -> float:
        if eest == 0:
            return 1 / self.qmax
        self.q11 = jpow(eest, self.beta1)
        q = jdiv(self.q11, jpow(self.errold, self.beta2))
        q /= self.gamma
        return jmin(1 / self.qmin, jmax(1 / self.qmax, q))

    def accept(self, eest: float, dt: float) -> float:
        q = self.q(eest)
        if q >= self.qsteady_min and q <= self.qsteady_max:
            q = 1
        self.errold = jmax(eest, self.qoldinit)
        return jdiv(dt, q)

    def reject(self, eest: float, dt: float) -> float:
        self.q11 = jpow(eest, self.beta1)
        return jdiv(dt, jmin(1 / self.qmin, self.q11 / self.gamma))

    def reset(self) -> None:
        self.errold = self.qoldinit
        self.q11 = 1.0


def initial_step(f: Callable[[float, np.ndarray], np.ndarray], t0: float, u0: np.ndarray, f0: np.ndarray,
                 tdir: float, order: float, reltol: float, abstol: Any, dtmax: float) -> float:
    """Hairer's starting step (Solving ODEs I, II.4), signed (``initialStep``)."""
    n = u0.size
    w = abstol + reltol * np.abs(u0)
    a = u0 / w
    b = f0 / w
    d0 = math.sqrt(seq_sum(a * a) / n)
    d1 = math.sqrt(seq_sum(b * b) / n)
    h0 = 1e-6 if (d0 < 1e-5 or d1 < 1e-5) else 0.01 * jdiv(d0, d1)
    h0 = jmin(h0, abs(dtmax))
    u1 = u0 + tdir * h0 * f0
    f1 = f(t0 + tdir * h0, u1)
    c = (f1 - f0) / w
    d2 = jdiv(math.sqrt(seq_sum(c * c) / n), h0)
    dmax = jmax(d1, d2)
    h1 = jmax(1e-6, h0 * 1e-3) if dmax <= 1e-15 else jpow(jdiv(0.01, dmax), 1 / (order + 1))
    h = jmin(jmin(100 * h0, h1), abs(dtmax))
    return tdir * h
